### PR TITLE
[IndexTable] Prevent sort icon flash and animation on right-aligned headers after row selection

### PR DESCRIPTION
**Pull request by rcaplanshopify**
### WHY are these changes introduced?

The recently introduced right-aligned, sortable header treatment flashed the hidden sort icon and applied the sort button slide animation when a row selection event occurred. These updates introduce logic to determine when a sortable, right-aligned header is rendering after a row selection event. If so, sort button and sort icon classes for right-aligned headers are not applied.

It seems as though the issue results from the `IndexSelectionChangeContext` updating and the resulting render of the `IndexTable`.

#### Before

https://user-images.githubusercontent.com/100380574/226065047-7ea83949-05a4-4384-bb74-101c14902d1e.mov

#### After

https://user-images.githubusercontent.com/100380574/226065057-96e0995d-9fa9-4ff8-8f9c-aed0450f66fe.mov

Fixes #0000 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/IndexTable/IndexTable.tsx
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.tsx:881]: Without the `hasSelected` logic, the flicker issue is mostly resolved, except for the first time a row or "select all" is selected.

File: polaris-react/src/components/IndexTable/IndexTable.tsx
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.tsx:928]: This line catches the first selection, whether it's the first click event or the first selection after the user interacts with the sortable headers.

File: polaris-react/src/components/IndexTable/IndexTable.tsx
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.tsx:948]: This prevents the flicker of the sort icon.

File: polaris-react/src/components/IndexTable/IndexTable.tsx
Comment (rcaplanshopify) [inline polaris-react/src/components/IndexTable/IndexTable.tsx:972]: This prevents the sort button from sliding after a right-aligned header is selected, then a different sort header is clicked, and then a single row checkbox or the "select all" checkbox is clicked.

**Review by alex-page — APPROVED** [review]
Happy to add this but adding new `useEffect` should be quite rare. I trust you tried a bunch of different solutions before landing here.

**Comment by mrcthms** [comment]
Have we removed the animation from the heading if it's right aligned? Or is it just some video capture funkiness in the "after" recording?

**Comment by rcaplanshopify** [comment]
@mrcthms I think it's a combination of video capture funkiness/frame rate and the reduced animation durations (compared to the initial iteration).

@alex-page Great call out on the `useEffect`. I removed it and everything still works as expected.

Here's a new video capture taken after the removal of the `useEffect` hook.

https://user-images.githubusercontent.com/100380574/226413869-b982491d-e773-44d4-9917-180190809018.mov

**Review by alex-page — APPROVED** [review]

**Comment by mrcthms** [comment]
@rcaplanshopify I'm not sure if this is directly related to the changes made in this PR, but I'm getting some funky behaviour when hovering over the area to the right of a right-aligned heading. I think we'd want to make sure we address this before merging

https://user-images.githubusercontent.com/2562596/226624044-f7f666d2-36ee-4841-a6f9-ab5947ebfb05.mov

**Comment by rcaplanshopify** [comment]
> @rcaplanshopify I'm not sure if this is directly related to the changes made in this PR, but I'm getting some funky behaviour when hovering over the area to the right of a right-aligned heading. I think we'd want to make sure we address this before merging

I think it is related. My guess is that hovering over the hidden sort icon causes the button containing it to slide to the left, which ends the hover event. I'll try to resolve this today. Thanks for pointing it out.

**Comment by rcaplanshopify** [comment]
@mrcthms @alex-page I was able to prevent the hover flicker by [expanding the button's area with the `before` pseudo-element](https://github.com/Shopify/polaris/pull/8707/commits/c0c96e12ba340b76bbd751301f0d85ac41756e77).
https://user-images.githubusercontent.com/100380574/226697526-75c54d76-e63c-4fed-b200-bff609ef240e.mov

**Review by mrcthms — APPROVED** [review]
Looking good! Probably shouldn't block the release of this as this fixes what it set out to, but in the most recent video at around 18 seconds or so, I see the sort icon appear for a split second to the right of the column? Doesn't happen every time, and the vast majority of the time it doesn't appear, but possibly something to investigate if you find yourself with some free time

**Pull request merged**
